File: biddo_client/lib/core/repositories/notification.dart
import 'package:get/get.dart';
import '../models/notification.dart';
import 'base.dart';

class NotificationRepository {
  var dio = Get.find<Api>();

  Future<int> getUnreadNotificationsCount() async {
    try {
      var response = await dio.api.get('/notification/unread');
      return response.data['unreadNotificationsCount'];
    } catch (error) {
      print('Coult not get unread notifications count: $error');
      return 0;
    }
  }

  Future<List<BiddoNotification>> loadForAccount([
    int page = 0,
    int perPage = 20,
  ]) async {
    try {
      var response = await dio.api.get('/notification/$page/$perPage');
      return List<BiddoNotification>.from(
        response.data.map(
          (el) => BiddoNotification.fromJSON(el),
        ),
      );
    } catch (error, stackTrace) {
      print('Coult not load notifications: $error, $stackTrace');
      return [];
    }
  }

  Future<bool> markAsRead(String id) async {
    try {
      await dio.api.put('/notification/$id');
      return true;
    } catch (error) {
      return false;
    }
  }

  Future<bool> markAllAsRead() async {
    try {
      await dio.api.put('/notification');
      return true;
    } catch (error) {
      return false;
    }
  }
}

File: biddo_web/src/core/repositories/notification.ts
import { BiddoNotification } from '../domain/notification'
import { RequestMaker, RequestType } from '../services/request-maker'

class NotificationRepository {
  private basePath = '/notification'

  public async getUnreadNotificationsCount(): Promise<number> {
    try {
      const response = (await RequestMaker.makeRequest({
        path: `${this.basePath}/unread`,
        method: RequestType.GET,
      })) as { unreadNotificationsCount: number }
      return response.unreadNotificationsCount
    } catch (error) {
      console.error('Could not get unread notifications count:', error)
      return 0
    }
  }

  public async loadForAccount(page = 0, perPage = 20): Promise<BiddoNotification[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        path: `${this.basePath}/${page}/${perPage}`,
        method: RequestType.GET,
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => BiddoNotification.fromJSON(el))
    } catch (error) {
      console.error('Could not load notifications:', error)
      return []
    }
  }

  public async markAsRead(id: string): Promise<boolean> {
    try {
      await RequestMaker.makeRequest({
        path: `${this.basePath}/${id}`,
        method: RequestType.PUT,
      })
      return true
    } catch (error) {
      console.error('Could not mark notification as read:', error)
      return false
    }
  }

  public async markAllAsRead(): Promise<boolean> {
    try {
      await RequestMaker.makeRequest({
        path: this.basePath,
        method: RequestType.PUT,
      })
      return true
    } catch (error) {
      console.error('Could not mark all notifications as read:', error)
      return false
    }
  }
}

const NotificationRepositoryInstance = new NotificationRepository()
export { NotificationRepositoryInstance as NotificationRepository }
